### ISSUE TITLE
exotic promise handling breaks context

### ISSUE THREAD

**Issue by gergelyke** [othiym23/async-listener#50]
Given the following code snippet:

```
const request = require('request-promise')
function makeRequest (options) {
  return new Promise((resolve, reject) => {
      request(options)
        .then((result) => {
          resolve(result)
        })
        .catch((err) => {
          reject(err)
        })
  })
}
```

Because of some extra logic I cannot simple `return request(options)` and have to do this.

But for some reason it breaks and the context is lost - any idea why this happens?

**Comment by gergelyke** [comment]
If you could point me into the right direction I can continue debugging it

**Comment by hayes** [comment]
I'll look I to this later today, I think I know what the issue is

**Comment by gergelyke** [comment]
thanks, looking forward to it!

**Comment by hayes** [comment]
Won't have time to look at it in detail until I get home tonight, but if
the return value of
https://github.com/othiym23/async-listener/blob/master/index.js#L328 is a
promise, the promise wrapper should return that promise rather than the one
returned when calling new Promise(wrappedExecutor)

**Comment by hayes** [comment]
I think I misunderstood the issue, I thought the bug was that promise constructors returning a promise did not work correctly, which is not actually something promises are supposed to do.

I believe that request-promise uses bluebird rather than native promises, bluebird has its own internal task queue, which can cause issues with async-listener (I have not tested this myself, but have heard that it is the case).

This may be out of scope for async-listener to handle itself

**Comment by hayes** [comment]
looking at bluebird, I am fairly confident that async-listener will not work correctly with bluebird promises. It may work in some cases, but as soon as you start having concurrent activity, it will start misbehaving in strange and hard to debug ways due to how tasks are scheduled. 

wrapping bluebird seems complicated from my brief exploration, but might be doable by wrapping:

_setFullfilled, _setRejected, and _settlePromises.
it would probably look something like:

```javascript

Promise.prototype._setFulfilled =  function () {
  this.__async_listener_propagator = wrap(function (fn, ctx, args) {
    return fn.apply(ctc, args)
  })

  return originalSetFulfilled.apply(this, arguments)
})

Promise.prototype._settlePromises =  function () {
  return this.__async_listener_propagator(originalSettlePromises, this, arguments)
})
```

I'm sure that is way over simplified and not sure if it would work, but if you wanted to make bluebird work, that is probably the general approach you would have to take (wrap being the thing that glue.js returns)

**Comment by Qard** [comment]
You should be able to just wrap the `_addCallbacks` function. That's what I do for traceview. See: https://github.com/appneta/node-traceview/blob/master/lib/probes/bluebird.js

**Comment by hayes** [comment]
I suppose that works in some cases. This is something that most
instrumentation does, but I consider to be incorrect. Continuations should
be from resolve to callbacks rather than from add callback to callback
execution, but this should work for 80% of use cases

**Comment by Qard** [comment]
Ah, yep. I see what you mean. Works for my uses though. It basically flattens the tree, which is actually what I want for my use. Seems like it'd work fine for this case too, since it's just trying to link the request callbacks back up to the request caller. Not really the "correct" way but... :neutral_face:

**Comment by ofrobots** [comment]
Another alternative might be to patch bluebird's `Promise._getDomain`.

**Comment by gergelyke** [comment]
Did some further investigation, with the native Promises as well.

Given the following snippet, the second expectation fails.

```javascript
ns.run(function () {
      ns.set(key, value)

      Promise.resolve()
      .then((res) => {
        expect(ns.get(key)).to.eql(value)
      })
      .catch(done)
      .then(function () {
        expect(ns.get(key)).to.eql(value)
        done()
      })
      .catch(done)
    })
```

However, if the first catch is removed it works like magic:

```javascript
ns.run(function () {
      ns.set(key, value)

      Promise.resolve()
      .then((res) => {
        expect(ns.get(key)).to.eql(value)
      })
       //.catch(done)
      .then(function () {
        expect(ns.get(key)).to.eql(value)
        done()
      })
      .catch(done)
    })
```

Any idea? :)

**Comment by hayes** [comment]
Yep, known and fixed issue, but apparently it was never published. I don't
have access to the npm package to cut a release
Cc @othiym23 @Qard

**Comment by Qard** [comment]
@hayes What's the status of this? Is there a PR for this I missed?

**Comment by hayes** [comment]
I believe this was fixed by #47 which has been published

**Comment by Qard** [comment]
A cool. Thanks for the update and closing this. 👍
